File: Providers/Mixpanel/AnalyticsKitMixpanelProvider.m
//
//  AnalyticsKitMixpanelProvider.m
//
//
//  Created by Zac Shenker on 5/11/12.
//  Copyright (c) 2012 Collusion. All rights reserved.
//

#import "Mixpanel.h"
#import "AnalyticsKitMixpanelProvider.h"

@implementation AnalyticsKitMixpanelProvider

-(id<AnalyticsKitProvider>)initWithAPIKey:(NSString *)apiKey {
    self = [super init];
    if (self) {
        [Mixpanel sharedInstanceWithToken:apiKey];
    }
    return self;
}

-(void)applicationWillEnterForeground {}
-(void)applicationDidEnterBackground {}
-(void)applicationWillTerminate {}

-(void)uncaughtException:(NSException *)exception {
    [[Mixpanel sharedInstance]track:@"Uncaught Exceptions" properties:
     [NSDictionary dictionaryWithObjectsAndKeys:
      [exception name], @"ename",
      [exception reason], @"reason",
      [exception userInfo], @"userInfo",
      nil]];
}

-(void)logScreen:(NSString *)screenName {
    [[Mixpanel sharedInstance] track:[@"Screen - " stringByAppendingString:screenName]];
}

-(void)logEvent:(NSString *)event {
    [self logEvent:event withProperties:nil timed:NO];
}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict {
    [self logEvent:event withProperties:dict timed:NO];
}

-(void)logEvent:(NSString *)event withProperty:(NSString *)key andValue:(NSString *)value {
    [self logEvent:event withProperties:[NSDictionary dictionaryWithObject:value forKey:key]];
}

- (void)logEvent:(NSString *)eventName timed:(BOOL)timed {
    [self logEvent:eventName withProperties:nil timed:timed];
}

- (void)logEvent:(NSString *)eventName withProperties:(NSDictionary *)dict timed:(BOOL)timed {
    if (timed) {
        [[Mixpanel sharedInstance] timeEvent:eventName];
    } else {
        [[Mixpanel sharedInstance] track:eventName properties:dict];
    }
}

-(void)endTimedEvent:(NSString *)eventName withProperties:(NSDictionary *)dict {
    // Mixpanel documentation: timeEvent followed by a track with the same event name would record the duration
    [[Mixpanel sharedInstance] track:eventName properties:dict];
}

-(void)logError:(NSString *)name message:(NSString *)message exception:(NSException *)exception {
    [[Mixpanel sharedInstance] track:@"Exceptions" properties:
     [NSDictionary dictionaryWithObjectsAndKeys:
      name, @"name",
      message, @"message",
      [exception name], @"ename",
      [exception reason], @"reason",
      [exception userInfo], @"userInfo",
      nil]];
}

-(void)logError:(NSString *)name message:(NSString *)message error:(NSError *)error {
    [[Mixpanel sharedInstance] track:@"Errors" properties:
     [NSDictionary dictionaryWithObjectsAndKeys:
      name, @"name",
      message, @"message",
      [error localizedDescription], @"description",
      [NSString stringWithFormat:@"%ld", (long)[error code]], @"code",
      [error domain], @"domain",
      [[error userInfo] description], @"userInfo",
      nil]];
}

@end

File: Providers/Mixpanel/AnalyticsKitMixpanelProvider.swift
import Foundation
import Mixpanel

public class AnalyticsKitMixpanelProvider: NSObject, AnalyticsKitProvider {

    @objc(initWithAPIKey:)
    public init(withAPIKey apiKey: String) {
        Mixpanel.sharedInstance(withToken: apiKey)
    }

    public func uncaughtException(_ exception: NSException) {
        Mixpanel.sharedInstance()?.track("Uncaught Exceptions", properties: [
            "ename" : exception.name,
            "reason" : exception.reason ?? "nil",
            "userInfo" : exception.userInfo ?? "nil"
        ])
    }

    // Logging

    public func applicationWillEnterForeground() {}
    public func applicationDidEnterBackground() {}
    public func applicationWillTerminate() {}

    public func logScreen(_ screenName: String) {
        logEvent("Screen - \(screenName)")
    }

    public func logScreen(_ screenName: String, withProperties properties: [String: Any]) {
        logEvent("Screen - \(screenName)", withProperties: properties)
    }

    public func logEvent(_ event: String) {
        Mixpanel.sharedInstance()?.track(event)
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any]) {
        Mixpanel.sharedInstance()?.track(event, properties: properties)
    }

    public func logEvent(_ event: String, timed: Bool) {
        if timed {
            Mixpanel.sharedInstance()?.timeEvent(event)
        } else {
            Mixpanel.sharedInstance()?.track(event)
        }
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any], timed: Bool) {
        if timed {
            Mixpanel.sharedInstance()?.timeEvent(event)
        } else {
            Mixpanel.sharedInstance()?.track(event, properties: properties)
        }
    }

    public func logEvent(_ event: String, withProperty key: String, andValue value: String) {
        logEvent(event, withProperties: [key: value])
    }

    public func endTimedEvent(_ event: String, withProperties properties: [String: Any]) {
        // Mixpanel documentation: timeEvent followed by a track with the same event name would record the duration
        Mixpanel.sharedInstance()?.track(event, properties: properties)
    }

    public func logError(_ name: String, message: String?, properties: [String: Any]?, exception: NSException?) {
        var loggedProperties = [String: Any]()
        loggedProperties["name"] = name
        loggedProperties["message"] = message
        if let exception = exception {
            loggedProperties["ename"] = exception.name
            loggedProperties["reason"] = exception.reason
            loggedProperties["userInfo"] = exception.userInfo
        }
        if let properties = properties {
            loggedProperties.merge(properties) { (current, _) in current }
        }

        Mixpanel.sharedInstance()?.track("Exceptions", properties: loggedProperties)
    }

    public func logError(_ name: String, message: String?, properties: [String: Any]?, error: Error?) {
        var loggedProperties = [String: Any]()
        loggedProperties["name"] = name
        loggedProperties["message"] = message
        if let error = error {
            loggedProperties["description"] = error.localizedDescription
        }
        if let properties = properties {
            loggedProperties.merge(properties) { (current, _) in current }
        }

        Mixpanel.sharedInstance()?.track("Errors", properties: loggedProperties)
    }
}
